### PR TITLE
Add AssertSameDifferentTypesRuleTest

### PR DESCRIPTION
**Pull request by mhujer**
I wanted to have a look on how to write PHPStan rules tests and I found none in this repository. So I created one.

### CODE REVIEW

File: tests/Rules/PHPUnit/data/assert-same.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+class FooTestCase extends \PHPUnit\Framework\TestCase
+{
+
+	public function testAssertSame()
+	{
+		$this->assertSame('1', 1);
+		$this->assertSame('1', new \stdClass());
+		$this->assertSame(1, $this->returnsString());
+		$this->assertSame('1', self::returnsInt());
```
Comment (ondrejmirtes) [inline tests/Rules/PHPUnit/data/assert-same.php]: There should also be a few examples of positive cases - comparing int/int and probably also comparing some unions against simple type and unions against unions.

File: tests/Rules/PHPUnit/data/assert-same.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare(strict_types = 1);
```
Comment (ondrejmirtes) [inline tests/Rules/PHPUnit/data/assert-same.php]: Please put declare on the same line as `<?php` :)

File: tests/Rules/PHPUnit/data/assert-same.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace PHPUnit;
```
Comment (ondrejmirtes) [inline tests/Rules/PHPUnit/data/assert-same.php]: Please use a different namespace, something like `namespace ExampleTestCase` is fine 😊

Comment (mhujer) [reply tests/Rules/PHPUnit/data/assert-same.php]: The intention was to make it consistent with https://github.com/phpstan/phpstan-strict-rules/blob/master/tests/Rules/StrictCalls/data/strict-calls.php

Comment (ondrejmirtes) [reply tests/Rules/PHPUnit/data/assert-same.php]: Yeah, but `PHPUnit` is a well known and used namespace :)

Comment (mhujer) [reply tests/Rules/PHPUnit/data/assert-same.php]: ok :-)

File: tests/Rules/PHPUnit/data/assert-same.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		$this->assertSame(1, 1);
+		$this->assertSame(['a'], ['a', 'b']);
+		$this->assertSame('1', '1');
+		$this->assertSame('1', '2');
+		$this->assertSame(new \stdClass(), new \stdClass());
+		$this->assertSame('1', $this->returnsString());
+		$this->assertSame(1, self::returnsInt());
+
+		// this breaks PHPStan
+		$this->assertSame(['a'], ['a', 1]);
```
Comment (mhujer) [inline tests/Rules/PHPUnit/data/assert-same.php]: @ondrejmirtes this breaks PHPStan

```
PHPStan\ShouldNotHappenException : Internal error.
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Type\TypeCombinator.php:23
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Type\ArrayType.php:28
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\Scope.php:575
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\Scope.php:323
 C:\devweb\htdocs\phpstan-phpunit\src\Rules\PHPUnit\AssertSameDifferentTypesRule.php:62
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\Analyser.php:162
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\NodeScopeResolver.php:316
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\NodeScopeResolver.php:176
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\NodeScopeResolver.php:686
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\NodeScopeResolver.php:176
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\NodeScopeResolver.php:686
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\NodeScopeResolver.php:176
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\NodeScopeResolver.php:686
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\NodeScopeResolver.php:176
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Analyser\Analyser.php:166
 C:\devweb\htdocs\phpstan-phpunit\vendor\phpstan\phpstan\src\Testing\RuleTestCase.php:64
 C:\devweb\htdocs\phpstan-phpunit\tests\Rules\PHPUnit\AssertSameDifferentTypesRuleTest.php:19
```

File: tests/Rules/PHPUnit/data/assert-same.php (hunk 1 above)
Comment (mhujer) [reply tests/Rules/PHPUnit/data/assert-same.php]: added bellow

File: tests/Rules/PHPUnit/data/assert-same.php (hunk 4 above)
Comment (mhujer) [reply tests/Rules/PHPUnit/data/assert-same.php]: fixed by enabling union types in tests bootstrap

**Comment by mhujer** [comment]
R4R @ondrejmirtes

**Comment by ondrejmirtes** [comment]
Awesome! :)

**Pull request merged**